### PR TITLE
feat: placeable keycaps

### PR DESCRIPTION
**Pull request by EvanGofourth**
Added code to make keycaps placeable. Created a Placeholder and a Placeable class. Placeholder will go on objects that represent where keycaps and switches would go. Placeable goes on a keycap/switch prefab. This should be easy to expand upon to give us all the functionality we need to place keycaps/switches/etc.

### CODE REVIEW

File: Assets/Scripts/SelectionSquare.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.EventSystems;
+
+public class SelectionSquare : MonoBehaviour, IPointerClickHandler
+{
+    public GameObject my_placement_object;
```
Comment (andrueandersoncs) [inline Assets/Scripts/SelectionSquare.cs]: We need to establish proper styles before we get too far into this. `camelCase` is the C# standard.

File: Assets/Scripts/SelectionSquare.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    public GameObject ui_handler;
+
+    public void Start()
+    {
+        ui_handler = GameObject.Find("UI_Manager");
```
Comment (andrueandersoncs) [inline Assets/Scripts/SelectionSquare.cs]: Game Object names can change, use a singleton MonoBehaviour if this is absolutely necessary and `FindObjectOfType<UIManager>()`

File: Assets/Scripts/SelectionSquare.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    }
+
+    public void OnPointerClick(PointerEventData eventData)
+    {
+        ui_handler.GetComponent<UIManager>().placement_object_prefab = my_placement_object;
```
Comment (andrueandersoncs) [inline Assets/Scripts/SelectionSquare.cs]: `GetComponent()` every time someone clicks? Come on. We're better than that. Call it once on `Awake()` and use a private reference.

File: Assets/Scripts/UIManager.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+public class UIManager : MonoBehaviour
+{
+    public GameObject placement_object_prefab;
+
+    Ray mouse_ray;
+    RaycastHit hit;
+
+    GameObject tile;
+
+    GraphicRaycaster m_Raycaster;
```
Comment (andrueandersoncs) [inline Assets/Scripts/UIManager.cs]: `m_Raycaster` ?? Big yikes my dude, it's obviously private.. no need to repeat yourself with `m_` - let's follow C# standards

File: Assets/Scripts/UIManager.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    void ClickManager()
+    {
+        if (Input.GetKeyDown(KeyCode.Mouse0))
+        {
+            if(focused_placeholder && placement_object_prefab)
+            {
+                // place the object prefab in the placeholder.
+                if(focused_placeholder.GetComponent<Placeholder>().type == placement_object_prefab.GetComponent<Placeable>().type)
```
Comment (andrueandersoncs) [inline Assets/Scripts/UIManager.cs]: 1. Use https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnMouseOver.html
2. Use https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnMouseDown.html
3. Let objects maintain their own behavior - if a placeholder is going to be changed, the component doing the changing should be on that object

File: Assets/Scripts/UIManager.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    GameObject tile;
+
+    GraphicRaycaster m_Raycaster;
+    PointerEventData m_PointerEventData;
+    EventSystem m_EventSystem;
+
+    public GameObject focused_placeholder;
+
+    private void Start()
```
Comment (andrueandersoncs) [inline Assets/Scripts/UIManager.cs:22]: remove unused MonoBehaviour messages, they will get called if they are defined thus slowing execution speed even if nothing is in the method

File: Assets/Scripts/UIManager.cs (hunk 5 above)
Comment (EvanGofourth) [reply Assets/Scripts/UIManager.cs]: What component is doing the 'changing' here though? Isn't it the user click? So I decided a UI Manager object should handle user click interactions. How would you propose item #3 look?

File: Assets/Scripts/UIManager.cs (hunk 4 above)
Comment (EvanGofourth) [reply Assets/Scripts/UIManager.cs]: fixed

File: Assets/Scripts/SelectionSquare.cs (hunk 3 above)
Comment (EvanGofourth) [reply Assets/Scripts/SelectionSquare.cs]: fixed

File: Assets/Scripts/SelectionSquare.cs (hunk 1 above)
Comment (EvanGofourth) [reply Assets/Scripts/SelectionSquare.cs]: fixed

File: Assets/Scripts/SelectionSquare.cs (hunk 2 above)
Comment (EvanGofourth) [reply Assets/Scripts/SelectionSquare.cs]: fixed

**Review by andrueandersoncs — APPROVED** [review]
Nice

**Pull request merged**